BitwardenShared/Core/Platform/Services/AuthenticatorSyncService.swift

```diff
@@ -0,0 +1,247 @@
+    ///
+    /// - Parameters:
+    ///   - ciphers: The encrypted `Cipher` objects.
+    ///   - userId: The userId of the account to which these Ciphers belong.
+    ///
+    /// - Returns: The decrypted, filtered, and sorted `CipherDTO` objects.
+    ///
+    private func decryptTOTPs(_ ciphers: [Cipher],
+                              userId: String) async throws -> [AuthenticatorBridgeItemDataView] {
+        let decryptedCiphers = try await ciphers.asyncMap { cipher in
+            try await self.clientService.vault().ciphers().decrypt(cipher: cipher)
+        }
+        let account = try await stateService.getActiveAccount()
+        let username = account.profile.name ?? account.profile.email
+
+        let ciphersToUse = decryptedCiphers.filter { cipher in
+            cipher.deletedDate == nil
+                && cipher.type == .login
+                && cipher.login?.totp != nil
+        }
+
+        return ciphersToUse.map { cipher in
+            AuthenticatorBridgeItemDataView(
+                favorite: false,
+                id: cipher.id ?? UUID().uuidString,
+                name: cipher.name,
+                totpKey: cipher.login?.totp,
+                username: username
+            )
+        }
+    }
+
+    /// This function handles the initial syncing with the Authenticator app as well as listening for updates
+    /// when the user adds new items. This is called when the sync is turned on.
+    ///
+    /// - Parameter userId: The userId of the user who has turned on sync.
+    ///
+    private func handleSyncOnForUserId(_ userId: String) async {
+        Logger.application.log("#### sync is on for userId: \(userId)")
+
+        if !vaultTimeoutService.isLocked(userId: userId) {
+            Logger.application.log("#### App in foreground and unlocked. Begin key creations.")
+            do {
+                try await createAuthenticatorKeyIfNeeded()
+            } catch {
+                errorReporter.log(error: error)
+            }
+            Logger.application.log("#### Subscribing to cipher updates")
+            subscribeToCipherUpdates(userId: userId)
+        }
+    }
+
+    /// This function handles stopping sync and cleaning up all sync-related items when a user has turned sync Off.
+    ///
+    /// - Parameter userId: The userId of the user who has turned off sync.
+    ///
+    private func handleSyncOffForUserId(_ userId: String) {
+        Logger.application.log("#### sync is off for userId: \(userId)")
+        Logger.application.log("#### Canceling cipher update subscription")
+        cipherPublisherTasks[userId]??.cancel()
+        cipherPublisherTasks[userId] = nil
+    }
+
+    /// Subscribe to NotificationCenter updates about if the app is in the foreground vs. background.
+    ///
+    private func subscribeToAppState() {
+        Task {
+            for await _ in notificationCenterService.willEnterForegroundPublisher() {
+                Logger.application.log("#### app entered foreground")
```

Comment:
🤔 Do we want the app to log this on PRD? I see its value on debugging but I wouldn't leave this for PRD.

Reply:
I was debating pulling them out entirely once we are close to done. They are really useful for now in debugging stuff though so I've moved them over to `.debug` in the latest. 👍